### PR TITLE
QA-patch tested against v2.35,v2.36,master

### PR DESCRIPTION
**Pull request by gto2023**
Patch adressing Issue #14 "QA-issues with sandbox-2.35". Testsuite runs ok with v2.35, v2.36 and the current master branch. At least the 2.3[56] versions are able to repeatedly install themselves with sandbox, usersandbox and network-sandbox options active in FEATURES.

### CODE REVIEW

File: libsandbox/canonicalize.c [hunk 1]
```diff
@@ -110,10 +110,17 @@ erealpath(const char *name, char *resolved)
...
 					break;
 				if (S_ISLNK(st.st_mode)) {
-					ssize_t cnt = readlink(rpath, rpath, path_max);
+					/*   avoid undefined behaviour resulting from passing rpath
+					 *   as source and destination buffer to readlink:
+					 *   warning: passing argument 2 to 'restrict'-qualified
+					 *   parameter aliases with argument 1 [-Wrestrict]
+					 */
+				        char buffer[path_max];
+					ssize_t cnt = readlink(rpath, buffer, sizeof(buffer));
```
Comment (floppym) [inline libsandbox/canonicalize.c:119]: This should probably be `sizeof(buffer) - 1` to leave space for the trailing null byte.

File: libsandbox/libsandbox.c [hunk 2]
```diff
@@ -131,7 +131,7 @@ int resolve_dirfd_path(int dirfd, const char *path, char *resolved_path,
 
 	save_errno();
 
-	size_t at_len = resolved_path_len - 1 - 1 - (path ? strlen(path) : 0);
+	/*unused: size_t at_len = resolved_path_len - 1 - 1 - (path ? strlen(path) : 0);*/
```
Comment (floppym) [inline libsandbox/libsandbox.c:134]: Please remove this line instead of commenting it out.

File: libsandbox/libsandbox.c [hunk 3]
```diff
@@ -141,7 +141,16 @@ int resolve_dirfd_path(int dirfd, const char *path, char *resolved_path,
...
+
+
+	/*   avoid undefined behaviour resulting from passing resolved_path 
+	 *   as source and destination buffer to readlink:
+	 *   C99 warning: passing argument 2 to 'restrict'-qualified 
+	 *   parameter aliases with argument 1 [-Wrestrict] 
+	 */
+	char buffer[resolved_path_len];
+
+	ssize_t ret = readlink(resolved_path, buffer, sizeof(buffer));
```
Comment (floppym) [inline libsandbox/libsandbox.c:153]: Again, we need to leave space for the trailing null byte.

File: src/sandbox.c [hunk 4]
```diff
@@ -252,7 +254,7 @@ int main(int argc, char **argv)
 		dputs("Setting up the required environment variables.");
 
 	/* If not in portage, cd into it work directory */
-	if ('\0' != sandbox_info.work_dir[0])
+	if (strlen(sandbox_info.work_dir))
```
Comment (floppym) [inline src/sandbox.c:257]: What is the reason for this change? The existing check is slightly faster.

File: src/sandbox.c [hunk 5]
```diff
@@ -34,10 +34,12 @@ const char sbio_fallback_path[] = "/dev/stderr";
 
 static int setup_sandbox(struct sandbox_info_t *sandbox_info, bool interactive)
 {
-	if (NULL != getenv(ENV_PORTAGE_TMPDIR)) {
-		/* Portage handle setting SANDBOX_WRITE itself. */
-		sandbox_info->work_dir[0] = '\0';
-	} else {
+        /* avoid using uninitialized fields */
+	memset(sandbox_info,0,sizeof(*sandbox_info));
```
Comment (floppym) [inline src/sandbox.c:38]: Instead of using memset, let's initialize the struct in `main`.
```
struct sandbox_info_t sandbox_info = {};
```

**Review by floppym — CHANGES_REQUESTED** [review]
I think this should really be split at least 2 commits: one to address the sandbox_info.work_dir warning, and the other to address the readlink warnings.

Please update the commit message(s) to something more meaningful. You don't need to mention how you have tested the change in the commit message.

Here is an example:
```
sandbox: fix compiler warning

src/environ.c:211:19: error: the comparison will always evaluate as ‘true’ for the address of ‘work_dir’ will never be NULL
```

**Comment by floppym** [comment]
Also, please review our copyright policy and add your Signed-off-by.

https://www.gentoo.org/glep/glep-0076.html

File: src/sandbox.c (hunk 4 above)
Comment (gto2023) [reply src/sandbox.c:257]: Nowadays strlen can be assumed to be implemented by a compiler builtin, and principally I'm  inclined to prefer readability over some minimal speed gain at the price of a barbed-wire look&feel, at least in code sections that run only once during initialization. But this assumption seems not be fully justified with sandbox, as the authors provide their own implementation of memory and string routines like strlen in C, for a rather sensible reason.

Comment (floppym) [reply src/sandbox.c:257]: I don't think sandbox re-implements `strlen`; where did you see that?

Readability is certain a valid reason to change this.

**Comment by gto2023** [comment]
You are right, my bad, it actually was strdup(const char *s) in libsandbox/memory.c, which allows a NULL ptr as parameter and subsequently returns NULL in this case. This could have made some sense in case of strlen, at least. I've already seen a bunch of such untidy constructs, only beginning with a widespread willy-nilly mix of signed and unsigned integer types (size_t, ssize_t and int), partly as a consequence of inconsiderate function definitions.

**Comment by gto2023** [comment]
Closing this before creating a new request which should fix the spotted weak points.
